### PR TITLE
Handle reductions with empty batches.

### PR DESCRIPTION
**Pull request by DrTodd13**
## Changes included in this PR

Fix for tpch q14.  If the reduction code sees an initial or subsequent batch with no rows then the arrow::Scalar generated for the reduction operation will be null and then null combined with anything afterward on subsequent batches will be null.  So, this PR basically ignores empty batches.

## Testing strategy

run_ci
tpch locally

## User facing changes

None

## Checklist
- [X] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [X] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md) 
- [X] I have installed + ran pre-commit hooks.

### CODE REVIEW

File: bodo/pandas/_physical_conv.cpp [hunk 1]
```diff
@@ -69,6 +69,11 @@ void PhysicalPlanBuilder::Visit(duckdb::LogicalFilter& op) {
     std::map<std::pair<duckdb::idx_t, duckdb::idx_t>, size_t> col_ref_map =
         getColRefMap(source_cols);
 
+    // Duckdb can produce empty filters.
+    if (op.expressions.size() == 0) {
+        return;
+    }
```
Comment (ehsantn) [inline bodo/pandas/_physical_conv.cpp:75]: Can we add a unit test for this case?

File: bodo/pandas/physical/reduce.h [hunk 2]
```diff
@@ -61,6 +61,15 @@ class PhysicalReduce : public PhysicalSource, public PhysicalSink {
...
                 output_scalars.push_back(out_scalar_batch);
+            } else if (!out_scalar_batch->is_valid) {
+                // If we get an empty batch which results in invalid
+                // arrow::Scalar result then just ignore it.
+                continue;
+            } else if (!output_scalars[i]->is_valid) {
+                // The last result can be null if there have been no rows
+                // seen thus far.  In which case, use the current result
+                // just on this batch as the result thus far.
+                output_scalars[i] = out_scalar_batch;
```
Comment (ehsantn) [inline bodo/pandas/physical/reduce.h:72]: Would be great to add unit tests for these cases (all empty reduction, first batch empty, later batches empty).

File: bodo/pandas/physical/project.h [hunk 3]
```diff
@@ -94,6 +94,18 @@ class PhysicalProjection : public PhysicalSourceSink {
                 } else {
                     col_names.emplace_back("comp");
                 }
+            } else if (expr->type == duckdb::ExpressionType::CONJUNCTION_AND ||
```
Comment (ehsantn) [inline bodo/pandas/physical/project.h:97]: Unit test would be good.

File: bodo/pandas/_physical_conv.cpp (hunk 1 above)
Comment (DrTodd13) [reply bodo/pandas/_physical_conv.cpp:75]: Done.

File: bodo/pandas/physical/project.h (hunk 3 above)
Comment (DrTodd13) [reply bodo/pandas/physical/project.h:97]: I think this one was where I was running some tpch query partially.  It doesn't generate this condition if you run the whole thing.  I think it was probably q14 but recreating this one might be too time consuming.

File: bodo/pandas/physical/sort.h [hunk 4]
```diff
@@ -84,7 +84,13 @@ class PhysicalSort : public PhysicalSource, public PhysicalSink {
                           std::vector<duckdb::ColumnBinding>& source_cols,
                           int64_t limit = -1, int64_t offset = -1)
         : PhysicalSort(logical_order.orders, input_schema, source_cols, limit,
-                       offset) {}
+                       offset) {
+        if (!logical_order.projection_map.empty()) {
```
Comment (ehsantn) [inline bodo/pandas/physical/sort.h:88]: Let' open an issue to implement this soon.

File: bodo/pandas/physical/reduce.h [hunk 5]
```diff
@@ -61,6 +61,17 @@ class PhysicalReduce : public PhysicalSource, public PhysicalSink {
             // Update reduction result
             if (iter == 0) {
                 output_scalars.push_back(out_scalar_batch);
+            } else if (!out_scalar_batch->is_valid) {
+                std::cout << "out_scalar_batch not valid" << std::endl;
```
Comment (ehsantn) [inline bodo/pandas/physical/reduce.h]:
```suggestion
```

File: bodo/pandas/physical/reduce.h [hunk 6]
```diff
@@ -61,6 +61,17 @@ class PhysicalReduce : public PhysicalSource, public PhysicalSink {
...
+                // If we get an empty batch which results in invalid
+                // arrow::Scalar result then just ignore it.
+                continue;
+            } else if (!output_scalars[i]->is_valid) {
+                std::cout << "output_scalars[i] not valid" << std::endl;
```
Comment (ehsantn) [inline bodo/pandas/physical/reduce.h]:
```suggestion
```

File: bodo/pandas/physical/join.h [hunk 7]
```diff
@@ -202,6 +211,12 @@ class PhysicalJoin : public PhysicalSourceSink, public PhysicalSink {
         // TODO[BSE-4820]: support joining on Indexes
         this->output_schema->metadata = std::make_shared<TableMetadata>(
             std::vector<std::string>({}), std::vector<std::string>({}));
+        if (this->output_schema->column_names.size() !=
+            logical_join.GetColumnBindings().size()) {
+            throw std::runtime_error(
+                "Join output schema has different number of columns than "
+                "LogicalFilter");
```
Comment (ehsantn) [inline bodo/pandas/physical/join.h]:
```suggestion
                "LogicalComparisonJoin");
```

File: bodo/pandas/physical/join.h [hunk 8]
```diff
@@ -73,8 +78,12 @@ class PhysicalJoin : public PhysicalSourceSink, public PhysicalSink {
                     cond.left->Cast<duckdb::BoundColumnRefExpression>();
                 auto& right_bce =
                     cond.right->Cast<duckdb::BoundColumnRefExpression>();
-                this->left_keys.push_back(left_bce.binding.column_index);
-                this->right_keys.push_back(right_bce.binding.column_index);
+                this->left_keys.push_back(
```
Comment (ehsantn) [inline bodo/pandas/physical/join.h:81]: Good catch.

**Review by ehsantn — APPROVED** [review]
Thanks @DrTodd13. Looks good to me.

**Review by scott-routledge2 — APPROVED** [review]
LGTM, thanks @DrTodd13

File: bodo/pandas/physical/reduce.h (hunk 2 above)
Comment (DrTodd13) [reply bodo/pandas/physical/reduce.h:72]: done
